text/0000-project-descriptor-images.md

```diff
@@ -0,0 +1,159 @@
+# Meta
+[meta]: #meta
+- Name: Support specifying images in project descriptor
+- Start Date: 2021-10-26
+- Author(s): [@phil9909](https://github.com/phil9909), [@modulo11](https://github.com/modulo11), [@pbusko](https://github.com/pbusko), [@jkutner](https://github.com/jkutner)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+Add a new array of tables `[[images]]` to the `project.toml` file.
+
+# Definitions
+[definitions]: #definitions
+
+* **project descriptor**: The `project.toml` file that describes a project.
+* **image**: the output of a running a buildpack(s) against a project
+* **image name**: fully qualified name of an OCI image in the form `<registry>/<name>:<tag>`.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, there is no support for specifying an output image in the `project.toml` although flags like `--tag` are already present.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Buildpack user will be able to specify the resulting images in the `project.toml`:
+
+# How it Works
+[how-it-works]: #how-it-works
+
+- Accepted project descriptor related RFCs
+  - https://github.com/buildpacks/rfcs/blob/main/text/0019-project-descriptor.md
+  - https://github.com/buildpacks/rfcs/blob/main/text/0054-project-descriptor-schema.md
+  - https://github.com/buildpacks/rfcs/blob/main/text/0080-builder-key-in-project-descriptor.md (could serve as a role-model)
+  - https://github.com/buildpacks/rfcs/blob/main/text/0084-project-descriptor-domains.md
+  - Closed PR which includes some similar changes: https://github.com/buildpacks/rfcs/blob/project-descriptor-simple/text/0000-project-descriptor.md
+
+Example:
+
+```toml
+[[images]]
+name = "spring-petclinic"
+registry = "gcr.io"
+tags = ["latest", "v1"]
+
+[[images]]
+name = "spring-petclinic"
+tags = ["latest", "v1"]
+
+[[images]]
+name = "spring-petclinic"
+registry = "private.registry.corp:8443"
+```
+
+This example will produce the following images:
+
+- `gcr.io/spring-petclinic:latest`
+- `gcr.io/spring-petclinic:v1`
+- `docker.io/spring-petclinic:latest`
+- `docker.io/spring-petclinic:v1`
+- `private.registry.corp:8443/spring-petclinic:latest`
```

Comment:
In working group we discussed how having multiple registries in a single build could be problematic if `pack` is running with `--publish`.

It may also create problems on re-build, because one registry might have the previous layers, and another may not. Thus, where does it pull the layers from (or not pull from)

Reply:
> In working group we discussed how having multiple registries in a single build could be problematic if `pack` is running with `--publish`.

Could you provide more details in how far this would be problematic?

> It may also create problems on re-build, because one registry might have the previous layers, and another may not. Thus, where does it pull the layers from (or not pull from)

Given we know the SHA sums of all layers, is it relevant from where `pack` would pull them?

Reply:
>> In working group we discussed how having multiple registries in a single build could be problematic if pack is running with --publish.

> Could you provide more details in how far this would be problematic?

The problem is that `--publish` causes pack to write directly to the remote registry (as opposed to writing locally and pushing an image). It's not clear _if_ this is possible with multiple remote registries. If it is not, it may mean the `--publish` option is incompatible with having images on multiple registries.

> Given we know the SHA sums of all layers, is it relevant from where pack would pull them?

I think the problem may occur when different registries have different layers for the previous build. The lifecycle would need to know how to determine which is the one to use. 

I think think this might be solvable by adding something like:

```toml
[[io.buildpacks.images.previous]]
name = "..."
registry = "..."
```

Reply:
> It's not clear if this is possible with multiple remote registries. If it is not, it may mean the --publish option is incompatible with having images on multiple registries.

Currently the lifecycle will throw an error during the analyze stage, if multiple registries are provided together with `--publish` flag.

Another option is to split the images blocks into two separate: main target and additional targets. The main target will be used as a previous image (similarly to what pack is doing by default)

Example:
```toml
[image.main]
name = "spring-petclinic"
registry = "gcr.io"
tag = "latest"

[[image.additional]]
registry = "private-corp:8080"
name = "spring-petclinic"
tags = ["v3", "v1"]
```

Reply:
We could also treat the first entry in the `[[images]]` table as the default.

Reply:
In my mind, I'm struggling to answer this because it greatly depends on whether we're trying to map it to `pack` arguments. 

If we were, @pbusko [suggestion](https://github.com/buildpacks/rfcs/pull/187#discussion_r764051852) makes the most sense. If we were not, either of @jkutner's [suggestion](https://github.com/buildpacks/rfcs/pull/187#discussion_r763041695) or @hone's [suggestion](https://github.com/buildpacks/rfcs/pull/187#discussion_r765160071) seem ideal to me.